# Практическая работа

# Задача

Один из способов повысить эффективность взаимодействия банка с клиентами — отправлять предложение о новой услуге не всем клиентам, а только некоторым, которые выбираются по принципу наибольшей склонности к отклику на это предложение.

Задача заключается в том, чтобы предложить алгоритм, который будет выдавать склонность клиента к положительному или отрицательному отклику на предложение банка. Предполагается, что, получив такие оценки для некоторого множества клиентов, банк обратится с предложением только к тем, от кого ожидается положительный отклик.


Для решения этой задачи загрузите файлы из базы в Postgres.
Эта БД хранит информацию о клиентах банка и их персональные данные, такие как пол, количество детей и другие.

Описание таблиц с данными представлено ниже.


**D_work**

Описание статусов относительно работы:
- ID — идентификатор социального статуса клиента относительно работы;
- COMMENT — расшифровка статуса.


**D_pens**

Описание статусов относительно пенсии:
- ID — идентификатор социального статуса;
- COMMENT — расшифровка статуса.


**D_clients**

Описание данных клиентов:
- ID — идентификатор записи;
- AGE	— возраст клиента;
- GENDER — пол клиента (1 — мужчина, 0 — женщина);
- EDUCATION — образование;
- MARITAL_STATUS — семейное положение;
- CHILD_TOTAL	— количество детей клиента;
- DEPENDANTS — количество иждивенцев клиента;
- SOCSTATUS_WORK_FL	— социальный статус клиента относительно работы (1 — работает, 0 — не работает);
- SOCSTATUS_PENS_FL	— социальный статус клиента относительно пенсии (1 — пенсионер, 0 — не пенсионер);
- REG_ADDRESS_PROVINCE — область регистрации клиента;
- FACT_ADDRESS_PROVINCE — область фактического пребывания клиента;
- POSTAL_ADDRESS_PROVINCE — почтовый адрес области;
- FL_PRESENCE_FL — наличие в собственности квартиры (1 — есть, 0 — нет);
- OWN_AUTO — количество автомобилей в собственности.


**D_agreement**

Таблица с зафиксированными откликами клиентов на предложения банка:
- AGREEMENT_RK — уникальный идентификатор объекта в выборке;
- ID_CLIENT — идентификатор клиента;
- TARGET — целевая переменная: отклик на маркетинговую кампанию (1 — отклик был зарегистрирован, 0 — отклика не было).
    
    
**D_job**

Описание информации о работе клиентов:
- GEN_INDUSTRY — отрасль работы клиента;
- GEN_TITLE — должность;
- JOB_DIR — направление деятельности внутри компании;
- WORK_TIME — время работы на текущем месте (в месяцах);
- ID_CLIENT — идентификатор клиента.


**D_salary**

Описание информации о заработной плате клиентов:
- ID_CLIENT — идентификатор клиента;
- FAMILY_INCOME — семейный доход (несколько категорий);
- PERSONAL_INCOME — личный доход клиента (в рублях).


**D_last_credit**

Информация о последнем займе клиента:
- ID_CLIENT — идентификатор клиента;
- CREDIT — сумма последнего кредита клиента (в рублях);
- TERM — срок кредита;
- FST_PAYMENT — первоначальный взнос (в рублях).


**D_loan**

Информация о кредитной истории клиента:
- ID_CLIENT — идентификатор клиента;
- ID_LOAN — идентификатор кредита.

**D_close_loan**

Информация о статусах кредита (ссуд):
- ID_LOAN — идентификатор кредита;
- CLOSED_FL — текущий статус кредита (1 — закрыт, 0 — не закрыт).

Ниже представлен минимальный список колонок, которые должны находиться в итоговом датасете после склейки и агрегации данных. По своему усмотрению вы можете добавить дополнительные к этим колонки.

    - AGREEMENT_RK — уникальный идентификатор объекта в выборке;
    - TARGET — целевая переменная: отклик на маркетинговую кампанию (1 — отклик был зарегистрирован, 0 — отклика не было);
    - AGE — возраст клиента;
    - SOCSTATUS_WORK_FL — социальный статус клиента относительно работы (1 — работает, 0 — не работает);
    - SOCSTATUS_PENS_FL — социальный статус клиента относительно пенсии (1 — пенсионер, 0 — не пенсионер);
    - GENDER — пол клиента (1 — мужчина, 0 — женщина);
    - CHILD_TOTAL — количество детей клиента;
    - DEPENDANTS — количество иждивенцев клиента;
    - PERSONAL_INCOME — личный доход клиента (в рублях);
    - LOAN_NUM_TOTAL — количество ссуд клиента;
    - LOAN_NUM_CLOSED — количество погашенных ссуд клиента.


Будьте внимательны при сборке датасета: это реальные банковские данные, в которых могут наблюдаться дубли, некорректно заполненные значения или значения, противоречащие друг другу. Для получения качественной модели необходимо предварительно очистить датасет от такой информации.

## Задание 1

В предыдущем задании вы собрали всю информацию о клиентах в одну таблицу, где одна строчка соответствует полной информации об одном клиенте.

Загрузите эту таблицу.

In [1]:
# your code here
import pandas as pd

In [2]:
df = pd.read_csv('datasets/bank_clients.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15594 entries, 0 to 15593
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   AGREEMENT_RK       15594 non-null  int64  
 1   AGE                15594 non-null  int64  
 2   SOCSTATUS_WORK_FL  15594 non-null  int64  
 3   SOCSTATUS_PENS_FL  15594 non-null  int64  
 4   GENDER             15594 non-null  int64  
 5   CHILD_TOTAL        15594 non-null  int64  
 6   DEPENDANTS         15594 non-null  int64  
 7   PERSONAL_INCOME    15594 non-null  float64
 8   LOAN_NUM_TOTAL     15594 non-null  int64  
 9   LOAN_NUM_CLOSED    15594 non-null  int64  
 10  TARGET             15594 non-null  int64  
dtypes: float64(1), int64(10)
memory usage: 1.3 MB


In [3]:
df.head()

AGREEMENT_RK  AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  GENDER  \
0      60099204   42                  1                  0       1   
1      62244665   28                  1                  0       1   
2      61050759   64                  1                  1       0   
3      62079659   54                  1                  0       1   
4      66583553   26                  1                  0       0   

   CHILD_TOTAL  DEPENDANTS  PERSONAL_INCOME  LOAN_NUM_TOTAL  LOAN_NUM_CLOSED  \
0            1           0          25000.0               1                1   
1            1           1          10000.0               1                1   
2            2           0          30000.0               1                1   
3            0           0          25000.0               1                1   
4            1           1          15000.0               2                2   

   TARGET  
0       1  
1       0  
2       0  
3       0  
4       0

Разбейте данные на тренировочную и тестовую часть в пропорции 80% к 20%, зафиксируйте `random_state = 42`.

In [4]:
# your code here
X = df.drop(['TARGET', 'AGREEMENT_RK'], axis=1)
y = df['TARGET']

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X , y, test_size=0.2, random_state=42)

На тренировочных данных обучите линейную модель классификации для предсказания целевой переменной (столбец `TARGET`).

Сделайте прогноз вероятности отклика на рекламную кампанию для тестовых данных.

In [7]:
# your code here
from sklearn.linear_model import LogisticRegression

model =LogisticRegression()
model.fit(X_train, y_train)
pred_test = model.predict(X_test) # предсказание класса

Переведите вероятности в классы по стандартному порогу (0.5) и на тестовом наборе данных вычислите метрики:

* accuracy
* precision
* recall
* f1-score

In [8]:
# your code here
pred_test[:25]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0])

In [9]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

accuracy_score(y_test, pred_test)

0.8788073100352677

In [10]:
y.value_counts(normalize=True)

TARGET
0    0.883801
1    0.116199
Name: proportion, dtype: float64

имеется дисбаланс классов

In [11]:
confusion_matrix(y_test, pred_test)

array([[2741,    2],
       [ 376,    0]])

### *Для 2741 клиентов, которые не откликнулись модель ошибочно предсказала двоих. 376 клиентов которые сотрудничали и откликнулись, модель записала в неоткликнувшиеся*

In [13]:
prob = model.predict_proba(X_test)
prob[:7]

array([[0.89354728, 0.10645272],
       [0.92097352, 0.07902648],
       [0.88826962, 0.11173038],
       [0.8602438 , 0.1397562 ],
       [0.85670925, 0.14329075],
       [0.78540853, 0.21459147],
       [0.83890955, 0.16109045]])

In [14]:
prob_churn = prob[:,1]
classes = prob_churn > 0.5

In [16]:
print(f"Accuracy: {accuracy_score(y_test, classes)}")
print(f"Precision: {precision_score(y_test, classes)}")
print(f"Recall: {recall_score(y_test, classes)}")
print(f"f1: {f1_score(y_test, classes)}")

Accuracy: 0.8788073100352677
Precision: 0.0
Recall: 0.0
f1: 0.0


In [17]:
confusion_matrix(y_test, classes)

array([[2741,    2],
       [ 376,    0]])

Целевая метрика для задачи - полнота, так как нам нужно найти максимум клиентов, кто может откликнуться на рекламу.

Но при этом точность не должна просесть, поэтому за ней тоже следим.

Разбейте тренировочные данные на `train` и `val` части в пропорции 3 к 1.

В цикле:

* переберите пороги от 0 до 1 с шагом 0.01
* вычислите для каждого порога значение метрик precision и recall
* подберите такой порог, при котором recall не меньше 0.66, а точность максимальна.

In [18]:
# your code here
df = pd.read_csv('datasets/bank_clients.csv')

X = df.drop(['TARGET', 'AGREEMENT_RK'], axis=1)
y = df['TARGET']

X_train, X_val, y_train, y_val = train_test_split(X , y, test_size=0.25, random_state=42)





In [19]:
max_precision = -1
border = 0
i = 0

model =LogisticRegression()
model.fit(X_train, y_train)
pred_test = model.predict(X_val)

prob = model.predict_proba(X_val)
prob_churn = prob[:,1]

while i <=1:
  classes = prob_churn > i
  if (recall_score(y_val, classes) >= 0.66) and (precision_score(y_val, classes) > max_precision):
    border = i
    max_precision = precision_score(y_val, classes)
  i += 0.01




In [20]:
border, max_precision

(0.09999999999999999, 0.14530310041647385)

Для выбранного порога посчитайте все метрики на тестовых данных. Сильно ли они отличаются от метрик на валидации?

In [22]:
# your code here
classes = prob_churn > border

print(f"Accuracy: {accuracy_score(y_val, classes)}")
print(f"Precision: {precision_score(y_val, classes)}")
print(f"Recall: {recall_score(y_val, classes)}")
print(f"f1: {f1_score(y_val, classes)}")


Accuracy: 0.4888432931520903
Precision: 0.14530310041647385
Recall: 0.6826086956521739
f1: 0.23960320488363218


In [23]:
confusion_matrix(y_val, classes)

array([[1592, 1847],
       [ 146,  314]])

*Модель выявила 314 клиентов с обратным откликом*

Выведите на экран в виде таблицы топ-6 признаков с наибольшими по модулю весами модели.

In [24]:
# your code here

weights = sorted(model.coef_[0], key=abs, reverse=True)
weights[:6]

[-0.05857324262600664,
 -0.0016170588322582896,
 -0.0016170588322582896,
 -0.001424406691465322,
 -0.0010526453001183963,
 -0.0008861125250719087]

In [25]:
weights = weights[:6]

In [26]:
coef_table = pd.DataFrame({'features': list(X_train.columns), 'weights': list(model.coef_[0])})
coef_table.sort_values(by='weights')

features   weights
0                AGE -0.058573
7     LOAN_NUM_TOTAL -0.001617
8    LOAN_NUM_CLOSED -0.001617
4        CHILD_TOTAL -0.001424
3             GENDER -0.001053
1  SOCSTATUS_WORK_FL -0.000886
5         DEPENDANTS -0.000536
2  SOCSTATUS_PENS_FL -0.000417
6    PERSONAL_INCOME  0.000018

## Задание 2

Добавьте в Streamlit-приложение визуализацию результатов модели:

* опцию выбора порога и вывод метрик качества в зависимости от выбранного порога

* вывод прогноза модели на выбранном объекте (клиенте) - вероятность отклика на рекламу.

In [27]:
import pickle

In [28]:
with open('model.pickle', 'wb') as f:
  pickle.dump(model, f)

## Бонус

Попробуйте применить другие модели классификации для решения этой задачи (любые какие знаете).

Удалось ли добиться улучшения качества модели?

In [ ]:
# your code here